github-auth-provider/main.go

```diff
@@ -0,0 +1,207 @@
+}
+
+func main() {
+	var opts Options
+	if err := env.LoadEnvForStruct(&opts); err != nil {
+		fmt.Printf("failed to load options: %v\n", err)
+		os.Exit(1)
+	}
+
+	legacyOpts := options.NewLegacyOptions()
+	legacyOpts.LegacyProvider.ProviderType = "github"
+	legacyOpts.LegacyProvider.ProviderName = "github"
+	legacyOpts.LegacyProvider.ClientID = opts.ClientID
+	legacyOpts.LegacyProvider.ClientSecret = opts.ClientSecret
+
+	// GitHub-specific options
+	if opts.GitHubTeams != nil {
+		legacyOpts.LegacyProvider.GitHubTeam = *opts.GitHubTeams
+	}
+	if opts.GitHubOrg != nil {
+		legacyOpts.LegacyProvider.GitHubOrg = *opts.GitHubOrg
+	}
+	if opts.GitHubRepo != nil {
+		legacyOpts.LegacyProvider.GitHubRepo = *opts.GitHubRepo
+	}
+	if opts.GitHubToken != nil {
+		legacyOpts.LegacyProvider.GitHubToken = *opts.GitHubToken
+	}
+	if opts.GitHubAllowUsers != nil {
+		legacyOpts.LegacyProvider.GitHubUsers = strings.Split(*opts.GitHubAllowUsers, ",")
+	}
+
+	oauthProxyOpts, err := legacyOpts.ToOptions()
+	if err != nil {
+		fmt.Printf("failed to convert legacy options to new options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxyOpts.Server.BindAddress = ""
+	oauthProxyOpts.MetricsServer.BindAddress = ""
+	oauthProxyOpts.Cookie.Refresh = time.Hour
+	oauthProxyOpts.Cookie.Name = "obot_access_token"
+	oauthProxyOpts.Cookie.Secret = opts.AuthCookieSecret
+	oauthProxyOpts.Cookie.Secure = strings.HasPrefix(opts.ObotServerURL, "https://")
+	oauthProxyOpts.UpstreamServers = options.UpstreamConfig{
+		Upstreams: []options.Upstream{
+			{
+				ID:            "default",
+				URI:           "http://localhost:8080/",
+				Path:          "(.*)",
+				RewriteTarget: "$1",
+			},
+		},
+	}
+	oauthProxyOpts.RawRedirectURL = opts.ObotServerURL + "/oauth2/callback"
+	oauthProxyOpts.ReverseProxy = true
+	if opts.AuthEmailDomains != "" {
+		oauthProxyOpts.EmailDomains = strings.Split(opts.AuthEmailDomains, ",")
+	}
+
+	if err = validation.Validate(oauthProxyOpts); err != nil {
+		fmt.Printf("failed to validate options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxy, err := oauth2proxy.NewOAuthProxy(oauthProxyOpts, oauth2proxy.NewValidator(oauthProxyOpts.EmailDomains, oauthProxyOpts.AuthenticatedEmailsFile))
+	if err != nil {
+		fmt.Printf("failed to create oauth2 proxy: %v\n", err)
+		os.Exit(1)
+	}
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "9999"
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/{$}", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(fmt.Sprintf("http://127.0.0.1:%s", port)))
+	})
+	mux.HandleFunc("/obot-get-state", obotGetState(oauthProxy))
+	mux.HandleFunc("/obot-get-icon-url", obotGetIconURL)
+	mux.HandleFunc("/", oauthProxy.ServeHTTP)
+
+	fmt.Printf("listening on 127.0.0.1:%s\n", port)
+	if err := http.ListenAndServe("127.0.0.1:"+port, mux); err != nil {
```

Comment:
`ListenAndServe` always returns an error. I think we should catch the "server closed" error and not `exit 1` in that case.

Reply:
done